src/fetch/object_id.rs

```diff
@@ -119,6 +124,44 @@ where
         }
     }
 
+    /// Fetches an activitypub object, either from local database (if possible), or over http.
+    pub async fn dereference_signed<A>(
+        &self,
+        data: &Data<<Kind as Object>::DataType>,
+        actor: &A,
+    ) -> Result<Kind, <Kind as Object>::Error>
+    where
+        <Kind as Object>::Error: From<Error> + From<anyhow::Error>,
+        A: Actor,
+    {
+        let db_object = self.dereference_from_db(data).await?;
+
+        // if its a local object, only fetch it from the database and not over http
+        if data.config.is_local_url(&self.0) {
+            return match db_object {
+                None => Err(Error::NotFound.into()),
+                Some(o) => Ok(o),
+            };
+        }
+
+        // object found in database
+        if let Some(object) = db_object {
+            // object is old and should be refetched
+            if let Some(last_refreshed_at) = object.last_refreshed_at() {
+                if should_refetch_object(last_refreshed_at) {
+                    return self
+                        .dereference_from_http_signed(data, Some(object), actor)
+                        .await;
+                }
+            }
+            Ok(object)
+        }
+        // object not found, need to fetch over http
+        else {
+            self.dereference_from_http_signed(data, None, actor).await
+        }
+    }
+
```

Comment:
Similar to above, this could be done with an internal method and two public wrappers `dereference()`, `dereference_signed()`.